src/audio-worklet/design-pattern/lib/ring-buffer.js

```diff
@@ -1,3 +1,250 @@
+   *
+   * @param  {number} newChannelCount The new channel count.
+   */
+  adaptChannel(newChannelCount) {
+    if (newChannelCount < this._maxChannelCount) {
+      this._channelCount = newChannelCount;
+    }
+  }
+
+  /**
+   * Getter for the buffer length in frames.
+   *
+   * @return {?number} Buffer length in frames.
+   */
+  get length() {
+    return this._isInitialized ? this._length : null;
+  }
+
+  /**
+   * Getter for the number of channels.
+   *
+   * @return {?number} Buffer length in frames.
+   */
+  get numberOfChannels() {
+    return this._isInitialized ? this._channelCount : null;
+  }
+
+  /**
+   * Getter for the maxixmum number of channels allowed for the instance.
+   *
+   * @return {?number} Buffer length in frames.
+   */
+  get maxChannelCount() {
+    return this._isInitialized ? this._maxChannelCount : null;
+  }
+
+  /**
+   * Returns a Float32Array object for a given channel index. If the channel
+   * index is undefined, it returns the reference to the entire array of channel
+   * data.
+   *
+   * @param  {number|undefined} channelIndex Channel index.
+   * @return {?Array} a channel data array or an
+   * array of channel data.
+   */
+  getChannelData(channelIndex) {
+    if (channelIndex >= this._channelCount) {
+      return null;
+    }
+
+    return typeof channelIndex === 'undefined' ?
+        this._channelData : this._channelData[channelIndex];
+  }
+
+  /**
+   * Returns the base address of the allocated memory space in the WASM heap.
+   *
+   * @return {number} WASM Heap address.
+   */
+  getHeapAddress() {
+    return this._dataPtr;
+  }
+
+  /**
+   * Returns the base address of the allocated memory space in the WASM heap.
+   *
+   * @return {number} WASM Heap address.
+   */
+  getPointer() {
+    return this._dataPtr;
+  }
+
+  /**
+   * Frees the allocated memory space in the WASM heap.
+   */
+  free() {
+    this._isInitialized = false;
+    this._module._free(this._dataPtr);
+    this._module._free(this._pointerArrayPtr);
+    this._channelData = null;
+  }
+
+/**
+ * A JS FIFO implementation for the AudioWorklet. 3 assumptions for the
+ * simpler operation:
+ *  1. the push and the pull operation are done by 128 frames. (Web Audio
+ *    API's render quantum size in the speficiation)
+ *  2. the channel count of input/output cannot be changed dynamically.
+ *    The AudioWorkletNode should be configured with the `.channelCount = k`
+ *    (where k is the channel count you want) and
+ *    `.channelCountMode = explicit`.
+ *  3. This is for the single-thread operation. (obviously)
+ *
+ *
+ */
+
+  /**
+   * Getter for Available frames in buffer.
+   *
+   * @return {number} Available frames in buffer.
+   */
+  get framesAvailable() {
+    return this._framesAvailable;
+  }
+
+  /**
+   * Push a sequence of Float32Arrays to buffer.
+   *
+   * @param  {array} arraySequence A sequence of Float32Arrays.
+   */
+  pushRing(arraySequence) {
```

Comment:
Not sure about the name of this method. I think we prefer "verb + noun", but here the noun part is unclear. What's "ring"? Being explicit always helps - `pushArraySequence`.

Also it looks like you're pushing `sample-frames` - a packet of multi-channel audio stream